Tpetra: Unable to use HostPinned, Host, or UVM for default buffer type for Cuda views

## Bug Report
@trilinos/tpetra
@mhoemmen @srajama1 @kyungjoo-kim

### Description
If you change the default 'Cuda' `buffer_type` from CudaSpace, you will get a build error in BlockCRS.  Doing a 'ninja -k 1000 Tptera_all` teases this down to a single error. The rest of Tpetra seems to work fine w/the change.

### Steps to Reproduce
Edit: `packages/tpetra/core/src/Tpetra_Details_DefaultTypes.hpp`
Change:
```
diff --git a/packages/tpetra/core/src/Tpetra_Details_DefaultTypes.hpp b/packages/tpetra/core/src/Tpetra_Details_DefaultTypes.hpp
index d1a2dd14cb..6b0a2fb403 100644
--- a/packages/tpetra/core/src/Tpetra_Details_DefaultTypes.hpp
+++ b/packages/tpetra/core/src/Tpetra_Details_DefaultTypes.hpp
@@ -128,7 +128,9 @@ namespace DefaultTypes {
 #ifdef KOKKOS_ENABLE_CUDA
     typename std::conditional<
       std::is_same<typename ExecutionSpace::execution_space, Kokkos::Cuda>::value,
-      Kokkos::CudaSpace,
+      // change this to something not CudaSpace
+      // Kokkos::CudaSpace,
+      Kokkos::CudaHostPinnedSpace,
       typename ExecutionSpace::memory_space>::type;
 #else
     typename ExecutionSpace::memory_space;
```

Get:
```
[1/1] Building CXX object packages/tpetra/core/src/CMakeFiles/tpetra.dir/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp.o
FAILED: packages/tpetra/core/src/CMakeFiles/tpetra.dir/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp.o 
/usr/tce/packages/spectrum-mpi/spectrum-mpi-rolling-release-gcc-7.3.1/bin/mpicxx  -Dtpetra_EXPORTS -I. -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/core/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/core/src/kokkos_refactor -Ipackages/tpetra/core/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/tsqr/src -Ipackages/tpetra/tsqr/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/classic/LinAlg -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/classic/NodeAPI -Ipackages/tpetra/classic/NodeAPI -Ipackages/tpetra/classic/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/classic/src -Ipackages/teuchos/kokkoscomm/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/kokkoscomm/src -Ipackages/teuchos/kokkoscompat/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/kokkoscompat/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/parameterlist/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/parser/src -Ipackages/teuchos/core/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/core/src -Ipackages/kokkos/core/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/core/src -I/usr/tce/packages/cuda/cuda-9.2.148/include -I/projects/trilinos/vortex/tpls/vortex_gcc-7.3.1_cuda-9.2.148_spectrum-rolling-release-a52c9c-10.2.0/boost-1.69.0/include -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/comm/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/remainder/src -Ipackages/teuchos/remainder/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/teuchos/numerics/src -Ipackages/kokkos-kernels/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/impl -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/impl/tpls -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/blas -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/blas/impl -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/sparse -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/sparse/impl -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/graph -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/graph/impl -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/batched -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/batched/impl -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos-kernels/src/common -Ipackages/kokkos/algorithms/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/algorithms/src -Ipackages/kokkos/containers/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/containers/src -Ipackages/epetra/src -I/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/epetra/src -fno-omit-frame-pointer -fstack-protector-all -fno-inline -g -mtune=power9 -mcpu=power9   --std=c++11 -lineinfo -g -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored -expt-extended-lambda -arch=sm_70   -g -O0 -fPIC -MD -MT packages/tpetra/core/src/CMakeFiles/tpetra.dir/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp.o -MF packages/tpetra/core/src/CMakeFiles/tpetra.dir/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp.o.d -o packages/tpetra/core/src/CMakeFiles/tpetra.dir/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp.o -c packages/tpetra/core/src/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp
/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp(2018): error: static assertion failed with "Incompatible View copy assignment"
          detected during:
            instantiation of "std::enable_if<Kokkos::Impl::ViewMapping<Kokkos::View<DataType, Properties...>::traits, Kokkos::View<RT, RP...>::traits, Kokkos::ViewTraits<DataType, Properties...>::specialize>::is_assignable_data_type, Kokkos::View<DataType, Properties...>>::type &Kokkos::View<DataType, Properties...>::operator=(const Kokkos::View<RT, RP...> &) [with DataType=char *, Properties=<Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Tpetra::Details::DefaultTypes::comm_buffer_memory_space<Kokkos::CudaUVMSpace::execution_space>>, void, void>, RT=char *, RP=<Kokkos::Cuda::array_layout, Kokkos::CudaUVMSpace::execution_space>]" 
/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/containers/src/Kokkos_DualView.hpp(708): here
            instantiation of "void Kokkos::DualView<DataType, Arg1Type, Arg2Type, Arg3Type>::resize(size_t, size_t, size_t, size_t, size_t, size_t, size_t, size_t) [with DataType=char *, Arg1Type=Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Tpetra::Details::DefaultTypes::comm_buffer_memory_space<Kokkos::CudaUVMSpace::execution_space>>, Arg2Type=void, Arg3Type=void]" 
/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp(2948): here
            instantiation of "void Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::packAndPrepareNew(const Tpetra::SrcDistObject &, const Kokkos::DualView<const Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::local_ordinal_type *, Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::buffer_device_type, void, void> &, Kokkos::DualView<char *, Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::buffer_device_type, void, void> &, Kokkos::DualView<size_t *, Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::buffer_device_type, void, void>, size_t &, Tpetra::Distributor &) [with Scalar=double, LO=int, GO=Tpetra::longlong, Node=Tpetra::Kokkos_Compat_KokkosCudaWrapperNode]" 
packages/tpetra/core/src/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp(76): here

/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/core/src/impl/Kokkos_ViewMapping.hpp(2973): error: static assertion failed with "View assignment must have compatible spaces"
          detected during:
            instantiation of "void Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::assign(Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::DstType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::SrcType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::TrackType &) [with DstTraits=Kokkos::ViewTraits<char *, Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Tpetra::Details::DefaultTypes::comm_buffer_memory_space<Kokkos::CudaUVMSpace::execution_space>>, void, void>, SrcTraits=Kokkos::ViewTraits<char *, Kokkos::Cuda::array_layout, Kokkos::CudaUVMSpace::execution_space>]" 
/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp(2019): here
            instantiation of "std::enable_if<Kokkos::Impl::ViewMapping<Kokkos::View<DataType, Properties...>::traits, Kokkos::View<RT, RP...>::traits, Kokkos::ViewTraits<DataType, Properties...>::specialize>::is_assignable_data_type, Kokkos::View<DataType, Properties...>>::type &Kokkos::View<DataType, Properties...>::operator=(const Kokkos::View<RT, RP...> &) [with DataType=char *, Properties=<Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Tpetra::Details::DefaultTypes::comm_buffer_memory_space<Kokkos::CudaUVMSpace::execution_space>>, void, void>, RT=char *, RP=<Kokkos::Cuda::array_layout, Kokkos::CudaUVMSpace::execution_space>]" 
/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/kokkos/containers/src/Kokkos_DualView.hpp(708): here
            instantiation of "void Kokkos::DualView<DataType, Arg1Type, Arg2Type, Arg3Type>::resize(size_t, size_t, size_t, size_t, size_t, size_t, size_t, size_t) [with DataType=char *, Arg1Type=Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Tpetra::Details::DefaultTypes::comm_buffer_memory_space<Kokkos::CudaUVMSpace::execution_space>>, Arg2Type=void, Arg3Type=void]" 
/projects/trilinos/jjellio/vortex/empire-work/Trilinos/packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp(2948): here
            instantiation of "void Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::packAndPrepareNew(const Tpetra::SrcDistObject &, const Kokkos::DualView<const Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::local_ordinal_type *, Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::buffer_device_type, void, void> &, Kokkos::DualView<char *, Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::buffer_device_type, void, void> &, Kokkos::DualView<size_t *, Tpetra::BlockCrsMatrix<Scalar, LO, GO, Node>::buffer_device_type, void, void>, size_t &, Tpetra::Distributor &) [with Scalar=double, LO=int, GO=Tpetra::longlong, Node=Tpetra::Kokkos_Compat_KokkosCudaWrapperNode]" 
packages/tpetra/core/src/Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp(76): here

2 errors detected in the compilation of "/tmp/tmpxft_000133b8_00000000-6_Tpetra_BlockCrsMatrix_DOUBLE_INT_LONG_LONG_CUDA.cpp1.ii".
ninja: build stopped: subcommand failed.
```

--- a/packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp
+++ b/packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp
@@ -2945,7 +2945,12 @@ public:
     // We use a "struct of arrays" approach to packing each row's
     // entries.  All the column indices (as global indices) go first,
     // then all their owning process ranks, and then the values.
-    exports.resize (totalNumBytes);
+    if(exports.extent(0) != totalNumBytes)
+    {
+      const std::string oldLabel = exports.d_view.label ();
+      const std::string newLabel = (oldLabel == "") ? "exports" : oldLabel;
+      exports = Kokkos::DualView<packet_type*, buffer_device_type>(newLabel, totalNumBytes);
+    }
     if (totalNumEntries > 0) {
       // Current position (in bytes) in the 'exports' output array.
       Kokkos::View<size_t*, host_exec> offset("offset", numExportLIDs+1);

